connector/src/main/java/org/jboss/as/connector/deployers/datasource/DataSourceDefinitionInjectionSource.java

```diff
@@ -355,6 +376,41 @@ private void setProperty(final DeploymentReflectionIndex deploymentReflectionInd
         }
     }
 
+    private static PathAddress getDataSourceAddress(final String jndiName, DeploymentUnit deploymentUnit, boolean xa) {
+        List<PathElement> elements = new ArrayList<>();
+        if (deploymentUnit.getParent() == null) {
+            elements.add(PathElement.pathElement(ModelDescriptionConstants.DEPLOYMENT, deploymentUnit.getName()));
+        } else {
+            elements.add(PathElement.pathElement(ModelDescriptionConstants.DEPLOYMENT, deploymentUnit.getParent().getName()));
+            elements.add(PathElement.pathElement(ModelDescriptionConstants.SUBDEPLOYMENT, deploymentUnit.getName()));
+        }
+        elements.add(PathElement.pathElement(ModelDescriptionConstants.SUBSYSTEM, DataSourcesExtension.SUBSYSTEM_NAME));
+        if (xa) {
+            elements.add(PathElement.pathElement(XA_DATA_SOURCE, jndiName));
+        } else {
+            elements.add(PathElement.pathElement(DATA_SOURCE, jndiName));
+        }
+        return PathAddress.pathAddress(elements);
+    }
+
+    private void installManagementModel(final CommonDataSource ds, final DeploymentUnit deploymentUnit, final PathAddress addr) {
+        final boolean xa = ds instanceof XaDataSource;
+
+        if (xa) {
+            XMLXaDataSourceRuntimeHandler.INSTANCE.registerDataSource(addr, (XaDataSource) ds);
+        } else {
+            XMLDataSourceRuntimeHandler.INSTANCE.registerDataSource(addr, (DataSource) ds);
+        }
+
+        final DeploymentResourceSupport deploymentResourceSupport = deploymentUnit.getAttachment(org.jboss.as.server.deployment.Attachments.DEPLOYMENT_RESOURCE_SUPPORT);
+        deploymentResourceSupport.getDeploymentSubModel(DataSourcesExtension.SUBSYSTEM_NAME, addr.getLastElement());
+
+        for (final Map.Entry<String, String> prop : properties.entrySet()) {
```

Comment:
I copied this weird loop from https://github.com/wildfly/wildfly/blob/master/connector/src/main/java/org/jboss/as/connector/deployers/ds/processors/DsXmlDeploymentInstallProcessor.java#L188
I'm not sure it is correct. Any thoughts?